File: Arbitration/MPExternalDisputeAPI/Model/ClaimCPT.cs
using System.Text.Json.Serialization;
using System.ComponentModel.DataAnnotations.Schema;
using Microsoft.EntityFrameworkCore;
using System.ComponentModel.DataAnnotations;

namespace MPExternalDisputeAPI.Model
{
    [Index(nameof(ArbitrationCaseId),nameof(CPTCode))]
    public class ClaimCPT
    {
        [JsonPropertyName("id")]
        public int Id { get; set; } = 0;

        // virtual properties
        [ForeignKey("ClaimCPTId")]
        [JsonPropertyName("caseSettlementCPTs")]
        public virtual List<CaseSettlementCPT> CaseSettlementCPTs { get; set; } = new List<CaseSettlementCPT>();
        
        //[ForeignKey("ClaimCPTId")]
        //[JsonPropertyName("disputeCPTs")]
        //public virtual List<AuthorityDisputeCPT> DisputeCPTs { get; set; } = new List<AuthorityDisputeCPT>();

        // properties
        [JsonPropertyName("cptCode")]
        [StringLength(20)]
        public string CPTCode { get; set; } = "";

        [JsonPropertyName("createdBy")]
        [StringLength(100)]
        public string CreatedBy { get; set; } = "";

        [JsonPropertyName("createdOn")]
        public DateTime? CreatedOn { get; set; } = null;

        // Fair Health CPT Data
        [JsonPropertyName("fh50thPercentileCharges")]
        public double FH50thPercentileCharges { get; set; } = 0;  // should be called Payor Allowed
        [JsonPropertyName("fh50thPercentileExtendedCharges")]
        public double FH50thPercentileExtendedCharges { get; set; } = 0;

        [JsonPropertyName("fh80thPercentileCharges")]
        public double FH80thPercentileCharges { get; set; } = 0; // should be called Provider Charges
        [JsonPropertyName("fh80thPercentileExtendedCharges")]
        public double FH80thPercentileExtendedCharges { get; set; } = 0; 

        [NotMapped]
        [JsonPropertyName("description")]
        [StringLength(300)]
        public string Description { get; set; } = "";  // ViewModel property used by the document templating utilities

        /** MPower field - allows soft delete and history tracking to continue */
        [JsonPropertyName("isDeleted")]
        public bool isDeleted { get; set; } = false; 
        [JsonPropertyName("isEligible")]
        public bool IsEligible { get; set; } = false;
        [JsonPropertyName("isIncluded")]
        public bool IsIncluded { get; set; } = false;
        [JsonPropertyName("isLocked")]
        public bool IsLocked { get; set; } = false;
        [JsonPropertyName("modifiers")]
        public string Modifiers { get; set; } = "";  // EHR flags
        [JsonPropertyName("modifier26_YN")]
        public bool Modifier26_YN { get; set; } = false;
        [JsonPropertyName("paidAmount")]
        public double PaidAmount { get; set; } = 0; // how much the payor has already paid on this CPT
        [JsonPropertyName("patientRespAmount")]
        public double PatientRespAmount { get; set; } = 0; // how much the patient is responsible form
        [JsonPropertyName("providerChargeAmount")]
        public double ProviderChargeAmount { get; set; } = 0;  // how much was originally charged for this CPT
        [JsonPropertyName("units")]
        public double Units { get; set; } = 0;
        [JsonPropertyName("updatedBy")]
        [StringLength(60)]
        public string UpdatedBy { get; set; } = "";
        [JsonPropertyName("updatedOn")]
        public DateTime? UpdatedOn { get; set; } = null;

        [JsonPropertyName("arbitrationCaseId")]
        public int ArbitrationCaseId { get; set; } = 0;
    }
}

File: Arbitration/MPArbitration/ClientApp/src/app/model/claim-cpt.ts
import { DisputeLinkVM } from './dispute-link-vm';
import { IModifier } from './imodifier';

export class ClaimCPT implements IModifier {
  public id = 0;
  public arbitrationCaseId = 0;
  public cptCode: string = '';
  public createdBy = '';
  public createdOn: Date | undefined;
  public disputes: DisputeLinkVM[] = [];

  // Fair Health CPT Data
  public fh50thPercentileCharges = 0; // i.e. Allowable Base amount
  public fh50thPercentileExtendedCharges = 0; // uses 50th only

  public fh80thPercentileCharges = 0; // i.e. 80th percentile base amount
  public fh80thPercentileExtendedCharges = 0;

  //public fhExtendedAdjustedAmount = 0;  // could apply a weight or cap to this but so far unused
  //public fhProviderFinalOfferNotToExceed = 0;  // capped by formula or app variable dependent on service line - not used at cpt level yet
  //public fhProviderFinalOffer= 0; // calculated offer using a formula - not used at cpt level yet

  /** MPower field - allows soft delete and history tracking to continue */
  public isDeleted = false;
  public isEligible = false; // eligible for arbitration
  public isIncluded = true;
  public isLocked = false; // prevents editing of this CPT

  public modifiers: string = ''; // EHR flags
  public modifier26_YN = false; // todo: move to a Viewmodel
  public paidAmount = 0; // how much the payor has already paid on this CPT
  public patientRespAmount = 0; // how much the patient is responsible form
  public providerChargeAmount = 0; // how much was originally charged for this CPT

  public units: number = 1;
  public updatedOn: Date | undefined;
  public updatedBy = '';

  constructor(obj?: any) {
    if (!obj) return;
    Object.assign(this, JSON.parse(JSON.stringify(obj)));
    if (this.createdOn) this.createdOn = new Date(obj.createdOn);
    if (this.updatedOn) this.updatedOn = new Date(obj.updatedOn);
  }
}

export class ClaimCPTBatchVM extends ClaimCPT {
  count = 0;
  constructor(obj: ClaimCPT) {
    super(obj);
  }
}
